src/error/Details.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for detailss.
+ */
+
+#ifndef SQUID_SRC_ERROR_DETAILS_H
+#define SQUID_SRC_ERROR_DETAILS_H
+
+#include "error/Detail.h"
+#include "mem/PoolingAllocator.h"
+
+#include <vector>
+
+/// Multiple details of a single error, in "canonical" order without duplicates.
+/// Our "canonical" order is the approximate detail discovery order (e.g., the
+/// order of Error::update() calls). This class isolates multi-detail storage
+/// overheads from a common case of storing a single error detail.
+class ErrorDetails: public ErrorDetail
+{
+public:
+    /// Combines the already stored and the latest error details while
+    /// preserving uniqueness and canonical order. Each parameter may point to a
+    /// single detail or to an ErrorDetails object (with multiple details).
+    static void Merge(ErrorDetailPointer &storage, const ErrorDetailPointer &latest);
+
+    virtual ~ErrorDetails() = default;
+
+protected:
+    // use Merge() instead
+    ErrorDetails() = default;
+
+    /// adds the given detail unless we already have it
+    void mergeOne(const ErrorDetail &);
+
+    /// adds unique details (if any) from the given collection
+    void mergeMany(const ErrorDetails &);
```

Comment:
This is **not** a self-documenting improvement. Both of these are performing the same semantic action (**merge**). The logic difference between type A and type B merging is already documented by the simple act of the merge being done on an A vs B object.

Reply:
FTR; at core meeting we decided to always present `ErrorDetails` or equivalent collection objects to callers to avoid the issues with juggling a mix of single instances and set containers through a single API.